### DISCUSSION TITLE
Implementing a feature of user-defined particle ionization from a user-input table

### DISCUSSION THREAD

**Discussion by GivenChen** [BLAST-WarpX/warpx#4332] [Ideas]
Hi,

Is it possible to implement a feature where users can use their own field ionization model for electron ionization?

It looks like only the ADK model is available in the field ionization method in WarpX, however, it is sometimes the case where the ADK model is incorrect and gives the wrong ionization probability in study. For example, right now I am simulating electron ionization in solid where the correct ionization model is the Keldysh model in solid, and the ADK model overestimates the ionization probability by many orders of magnitude. Additionally, for people using very high-frequency lasers, the ADK ionization model will also fail, and the multiphoton ionization formula is better to be used.

Right now most of the pic code uses the ADK ionization model, and not many codes offer other options. The only code I know is that PIConGPU offers BSI ionization and Keldysh ionization in gas. It would be really useful and helpful if Warpx could offer more options or even just offer a feature that allows users to prescribe their interested ionization model. 

For example, the user can write a table that gives the ionization rate at each instantaneous electric field as an additional input file. When running, Warpx code can calculate the ionization rate by interpolation of data from the input table. 

Replies would be highly appreciated!

Thank you

**Comment by ax3l** [comment]
Thank you for raising this.

We discussed this and should just implement the Keldysh ionization model in WarpX alongside ADK. Tracking added in issue #4335
Do you like to work on this? Otherwise we will schedule it to be worked on by one of us.

For more sophisticated models or custom/empiric implementations (e.g., with tables from third party rates/codes), we could expose you our ionization routines to Python. Would that help?

**Reply by GivenChen** [reply]
Hi, I have been thinking about this and I think I will leave it to more professional people to implement this Keldysh model. It looks like Keldysh model use instantaneous electric amplitude which is not a trivial problem for me... I also wonder do you plan to implement Keldysh ionization for solid, or only for gas?

Yeah, it would be really helpful if I can know the ionization routines to Python. I am not very familiar with using WarpX with Python and might ask more questions in this channel when I try to add custom ionization.

**Reply by ax3l** [reply]
> I also wonder do you plan to implement Keldysh ionization for solid, or only for gas?

Good question, is there a different one for those? Feel free to share the reference in #4335 and we can take it on.

> Yeah, it would be really helpful if I can know the ionization routines to Python. I am not very familiar with using WarpX with Python and might ask more questions in this channel when I try to add custom ionization.

Sounds good! Yes we definitely still have to expose ionization routine helpers to Python, but you can generally already grep particles from the PICMI examples and manipulate them at your will.

**Reply by ax3l** [reply]
HI @GivenChen, just sharing that we improved the docs on how to hook custom Python code into WarpX:
https://warpx.readthedocs.io/en/latest/usage/workflows/python_extend.html

**Comment by GivenChen** [comment]
Hi Axel,

Thank you for sharing this!
It is really nice to see this in documentation!

Best,
Given

Get Outlook for iOS<https://aka.ms/o0ukef>
________________________________
